Speech input

nmohamed opened the pull request:
Takes your words and turns it into an arraylist of possible things you said

nmohamed committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into speech_input

nmohamed committed:
created speechListener

formatting changes, added listener for literally listening for sound
input, added blurb in the fragment.
note: code won't work because the fragment seems to not be able to get
context.

nmohamed committed:
gets voice - need to return data

added button that starts recording voice
speech listener records voice and returns what you said

nmohamed committed:
added callback to get results

speech stuff is working

nmohamed committed:
merge conflict

greneroom commented:
There's merge conflicts...

nmohamed commented:
yeah im working on fixing them :cry:

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/MainFragment.java:
```diff
@@ -1,25 +1,77 @@
-        return inflater.inflate(R.layout.fragment_main, container, false);
+                             Bundle savedInstanceState) {
+        view = inflater.inflate(R.layout.fragment_main, container, false);
+        Button speechButton = (Button) view.findViewById(R.id.speech);
```
use [ButterKnife](http://jakewharton.github.io/butterknife/)

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/MainFragment.java:
```diff
@@ -1,25 +1,77 @@
+        listener = new SpeechListener(new SpeechCalback() {
+            @Override
+            public void callback(ArrayList voiceResult) {
+                Log.d(TAG, "callbacking -> you got resultzzz");
```
probably best to not have Debug messages like this on Master

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/MainFragment.java:
```diff
@@ -1,25 +1,77 @@
+            public void onClick(View v) {
+                if (!isListening) {
+                    doListen();
+                } else{
```
probably stop listening?

nmohamed replied:
it stops listening by itself, so it's unnecessary

davidabrahams replied:
but, maybe if we press the "listen" button while it's already listening the user wants it to stop listening? Siri cancels the input if you hit the button again

nmohamed replied:
okay, it will be done

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/SpeechListener.java:
```diff
@@ -0,0 +1,82 @@
+ * Voice Recognition: For speech to text input
+ */
+
+class SpeechListener implements RecognitionListener {
```
declare classes as public

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/SpeechListener.java:
```diff
@@ -0,0 +1,82 @@
+ */
+
+class SpeechListener implements RecognitionListener {
+    String TAG = "myDebug";
```
perhaps make this tag more specific?

nmohamed replied:
i left it as such for filtering debug comments so i only see those ones

davidabrahams replied:
I meant that maybe you should make the tag itself more specific, so if there's a debug statement in another Fragment you can tell the difference. Ie,

```
String SPEECH_TAG = "speechDebug"
```

nmohamed replied:
i did have it like that before but i couldn't find a way to filter
properly. ill change it though

On Wed, Nov 11, 2015 at 11:15 AM, greneroom notifications@github.com
wrote:

> In app/src/main/java/com/mobileproto/hireddit/hireddit/SpeechListener.java
> https://github.com/greneroom/Hi-Reddit/pull/7#discussion_r44550405:
> 
> > +import android.app.Fragment;
> > +import android.os.Bundle;
> > +import android.speech.RecognitionListener;
> > +import android.speech.SpeechRecognizer;
> > +import android.util.Log;
> > +import android.view.View;
> > +import android.widget.Toast;
> > +
> > +import java.util.ArrayList;
> > +
> > +/**
> > - \* Voice Recognition: For speech to text input
> > - */
> >   +
> >   +class SpeechListener implements RecognitionListener {
> > -    String TAG = "myDebug";
> 
> I meant that maybe you should make the tag itself more specific, so if
> there's a debug statement in another Fragment you can tell the difference.
> Ie,
> 
> String SPEECH_TAG = "speechDebug"
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/greneroom/Hi-Reddit/pull/7/files#r44550405.

greneroom commented:
Could you perhaps make your own Fragment and put the listener in it? That way, when I merge the `ui` branch, all your code doesn't get overwritten by the Fragments with UI components in them. Then we can take your code and attach it to the correct parts of the ui-fragments.

nmohamed commented:
so just copy and past everything to a new fragment?

greneroom commented:
Ya and leave the MainFragment as-is

On Wed, Nov 11, 2015 at 9:18 AM Nora Mohamed <notifications@github.com>
wrote:

> so just copy and past everything to a new fragment?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greneroom/Hi-Reddit/pull/7#issuecomment-155662428>.
>

greneroom commented:
And test that fragment obviously

On Wed, Nov 11, 2015 at 9:21 AM David Abrahams <
david.abrahams@students.olin.edu> wrote:

> Ya and leave the MainFragment as-is
>
> On Wed, Nov 11, 2015 at 9:18 AM Nora Mohamed <notifications@github.com>
> wrote:
>
>> so just copy and past everything to a new fragment?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/greneroom/Hi-Reddit/pull/7#issuecomment-155662428>.
>>
>

nmohamed committed:
made new fragment, changed code as requested

added butterknife, moved everything in main fragment to speech fragment,
made new xml for speech fragment.

*note that content_main.xml points to speech fragment, not main fragment

nmohamed committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into speech_input

Conflicts:
	app/src/main/java/com/mobileproto/hireddit/hireddit/MainActivity.java

LucyWilcox commented:
:last_quarter_moon_with_face: :yellow_heart:

nmohamed committed:
changed debug names

``Hi Reddit Debug"

nmohamed committed:
changed TAG again

tag now more descriptve

nmohamed committed:
added error toasts

lets the user know whats up.
*how detailed should error stuff be? i only put in details for errors
caused by the user

nmohamed commented:
had another idea for a stretch goal: a translated version! :es: :fr: :de: :gb: 
or if we want to be funny, slang version. idk

nmohamed committed:
shows partial results, added textview

still has issue where itll cut you off?

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/SpeechFragment.java:
```diff
@@ -0,0 +1,101 @@
+
+            @Override
+            public void errorCallback(int errorCode) {
+                if (errorCode == 7){
```
Could you add comments or link to documentation about what these error codes are?

nmohamed replied:
it's http://developer.android.com/reference/android/speech/SpeechRecognizer.html
i'll add a comment.

davidabrahams replied:
OHHHH I SEE. Okay, you defnitely shouldn't be hard coding in numbers. What you should do is:

```
if (errorCode == SpeechRecognizer.ERROR_NO_MATCH)
{
// handle the error
}
```

Do this for each of the errors.

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/SpeechFragment.java:
```diff
@@ -0,0 +1,101 @@
+                    doListen();
+                } else{
+                    //what you want to happen if you press the button an you're already listening for voice
+                    dontListen(); //note that you don't need to press the button again to stop listening - it'll automatically stop
```
nice :+1:

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/SpeechFragment.java:
```diff
@@ -0,0 +1,101 @@
+ * Speech Fragment: Holds functionality for requesting and recieving voice input
+ */
+public class SpeechFragment extends Fragment{
+    String TAG = "SpeechFragment Debug";
```
Change this variable name to `DEBUG_TAG`, just in case we want more tags.

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/SpeechListener.java:
```diff
@@ -0,0 +1,79 @@
+ */
+
+public class SpeechListener implements RecognitionListener {
+    String TAG = "SpeechListener Debug";
```
Same thing here, change `TAG`

davidabrahams commented on app/src/main/java/com/mobileproto/hireddit/hireddit/SpeechFragment.java:
```diff
@@ -0,0 +1,101 @@
+        ButterKnife.bind(this, view);
+
+        // *~Speech stuff~* //
+        listener = new SpeechListener(new SpeechCallback() {
```
we might want to subclass this, but for now this is fine

greneroom commented:
Address the few small changes I noted, then it's fine by me to merge!

LucyWilcox commented:
Looks nice! :+1: David for looking really closely and catching stuff..

Commit:
David's small stylistic changes
